src/librbd/api/Mirror.cc

```diff
@@ -3083,43 +3035,27 @@ int Mirror<I>::group_promote(IoCtx& group_ioctx, const char *group_name,
     return r;
   }
 
+  mirror::PromotionState promotion_state;
   cls::rbd::MirrorGroup mirror_group;
-  r = cls_client::mirror_group_get(&group_ioctx, group_id, &mirror_group);
+  C_SaferCond ctx;
+  // pass group_id to get ENOENT error returned back when group is not mirror enabled
+  auto request = mirror::GroupGetInfoRequest<I>::create(
+    group_ioctx, "", group_id, &mirror_group, &promotion_state, &ctx);
+  request->send();
+  r = ctx.wait();
   if (r == -ENOENT) {
     ldout(cct, 10) << "mirroring for group " << group_name
-                   << " disabled" << dendl;
-    return -EINVAL;
+                   << " already disabled" << dendl;
+    return 0;
   } else if (r < 0) {
-    lderr(cct) << "failed to retrieve mirror group metadata: "
-               << cpp_strerror(r) << dendl;
-    return r;
-  } else if (mirror_group.mirror_image_mode !=
-             cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT) {
-    auto mode = static_cast<rbd_mirror_image_mode_t>(
-        mirror_group.mirror_image_mode);
-    lderr(cct) << "cannot promote, mirror mode is set to: " << mode << dendl;
-    return -EOPNOTSUPP;
-  }
-
-  cls::rbd::MirrorSnapshotState state;
-  cls::rbd::GroupSnapshotState sync;
-  r = get_last_mirror_snapshot_state(group_ioctx, group_id, &state, &sync);
-  if (r == -ENOENT) {
-    sync = cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE;
-    state = cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED; // XXXMG?
-    r = 0;
-  }
-  if (r < 0) {
-    lderr(cct) << "failed to get last mirror snapshot state: "
-               << cpp_strerror(r) << dendl;
+    lderr(cct) << "failed to get group info: " << cpp_strerror(r) << dendl;
     return r;
   }
 
-  if (state == cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
+  if (promotion_state == mirror::PROMOTION_STATE_PRIMARY) {
     lderr(cct) << "group " << group_name << " is already primary" << dendl;
     return -EINVAL;
-  } else if (!force && (state == cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY ||
-                        sync != cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE)) {
+  } else if (!force && promotion_state == mirror::PROMOTION_STATE_NON_PRIMARY) {
```

Comment:
This will proceed to promote the group if the last snap is an incomplete non-primary demoted snap.

Reply:
Thinking of a last failed promote request, We should allow the next one,  correct?

Reply:
This is not the scenario of a failed promote request, it is a demote snap not having been synced fully .

Reply:
Ah! Right. I realize it the comparison to "==" now. 

The the promote should fail seeing the incomplete snap ? Unless a force promote is issued ?

Reply:
Yes, I think existing logic should be preserved here.  This case is special because it's the only place that relied on `sync` as returned by `get_last_mirror_snapshot_state()`.

Reply:
Done. Thanks